app/modules/sntp.c

```diff
@@ -698,8 +698,16 @@ static char *state_init(lua_State *L) {
 
 static char *set_repeat_mode(lua_State *L, bool enable)
 {
+  if (repeat) {
+    os_timer_disarm (&repeat->timer);
+    luaL_unref (L, LUA_REGISTRYINDEX, repeat->sync_cb_ref);
+    luaL_unref (L, LUA_REGISTRYINDEX, repeat->err_cb_ref);
+    luaL_unref (L, LUA_REGISTRYINDEX, repeat->list_ref);
+    c_free(repeat);
+    repeat = NULL;
+  }
+
   if (enable) {
-    set_repeat_mode(L, FALSE);
     repeat = (sntp_repeat_t *) c_malloc(sizeof(sntp_repeat_t));
```

Comment:
We should probably take the opportunity to change this to `repeat = luaM_new(L, sntp_repeat_t);` while we're at it. And of course, the `c_free(repeat)`  above becomes a `luaM_free(L, repeat);`

Reply:
Can do, but... what's the advantage, sorry?

Reply:
By using the Lua memory allocator we tap into its ability to automatically do a gc sweep if needed, reducing the likelihood of getting an out-of-memory in the first place, and if we do then the LVM will raise the error for us, so we don't need explicit null checking in our code.

Reply:
> And of course, the `c_free(repeat)` above becomes a `luaM_free(L, repeat);`

`luaM_freearray(L, repeat, sizeof(sntp_repeat_t));` to be precise :wink:
Ref. https://github.com/nodemcu/nodemcu-firmware/issues/2377#issuecomment-391135540

Reply:
@devsaurus Why? `repeat` has the correct type, so the resulting `sizeof(*repeat)` will yield the correct size in `luaM_free()`. It's only if you're casting things to a `char *` or are actually using arrays that you need the `luaM_freearray()` (or `luaM_freemem()`). Check `lmem.h` if you don't believe me! :)

Reply:
I haven't been tracking this module or this dialog so I can't make any detailed review comments.  I do that one general observation though: switching from the malloc allocator to the luaM allocator doesn't really impact leakage management either way.  The latter helps the Lua allocator to track used storage, but not GC.  

If you want the Lua GC to manage resource clean up then you need to use collectable objects such as Userdata (not Light userdata which is not collectable).  Userdata can have a bound metatable with a `__GC` method defined so that you can carry out any cleanup needed if the GC was determined that the object is no longer referenced.